cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/DefaultArtifactManager.java

```diff
@@ -0,0 +1,197 @@
+    }
+
+    if (isSuccessful(httpResponse.getResponseCode())) {
+      List<ArtifactInfo> artifactInfoList =
+        GSON.fromJson(httpResponse.getResponseBodyAsString(), new TypeToken<List<ArtifactInfo>>() { }.getType());
+      return artifactInfoList;
+    } else {
+      throw new IOException(String.format("Exception while getting artifacts list %s",
+                                          httpResponse.getResponseBodyAsString()));
+    }
+  }
+
+  private boolean isSuccessful(int responseCode) {
+    return responseCode == 200;
+  }
+
+  private URL discoverServiceAndCreateURL(String pathSuffix) throws IOException {
+    ServiceDiscovered discovered = discoveryServiceClient.discover(Constants.Service.APP_FABRIC_HTTP);
+    Discoverable discoverable = new RandomEndpointStrategy(discovered).pick(1, TimeUnit.SECONDS);
+
+    if (discoverable == null) {
+      throw new IOException(String.format("Could not discover %s Service", Constants.Service.APP_FABRIC_HTTP));
+    }
+
+    InetSocketAddress address = discoverable.getSocketAddress();
+    String scheme = Arrays.equals(Constants.Security.SSL_URI_SCHEME.getBytes(), discoverable.getPayload()) ?
+      Constants.Security.SSL_URI_SCHEME : Constants.Security.URI_SCHEME;
+
+    String path = String.format("%s%s:%d%s/%s", scheme, address.getHostName(), address.getPort(),
+                                Constants.Gateway.API_VERSION_3, pathSuffix);
+
+    return new URL(path);
+  }
+
+  /**
+   * Create a class loader with artifact jar unpacked contents and parent for this classloader is the supplied
+   * parentClassLoader, if that parent classloader is null, bootstrap classloader is used as parent.
+   * This is a closeable classloader, caller should call close when he is done using it, during close directory
+   * cleanup will be performed.
+   *
+   * @param namespaceId artifact namespace
+   * @param artifactInfo artifact info whose artiact will be unpacked to create classloader
+   * @param parentClassLoader  optional parent classloader, if null bootstrap classloader will be used
+   * @return CloseableClassLoader call close on this CloseableClassLoader for cleanup
+   * @throws IOException if artifact is not found or there were any error while getting artifact
+   */
+  public CloseableClassLoader createClassLoader(
+    NamespaceId namespaceId, ArtifactInfo artifactInfo, @Nullable ClassLoader parentClassLoader) throws IOException {
+    String namespace = ArtifactScope.SYSTEM.equals(artifactInfo.getScope()) ?
+      NamespaceId.SYSTEM.getNamespace() : namespaceId.getEntityName();
+
+    URL url = discoverServiceAndCreateURL(
+      String.format("namespaces/%s/artifact-internals/artifacts/%s/versions/%s/location",
+                    namespace, artifactInfo.getName(), artifactInfo.getVersion()));
+
+    HttpResponse httpResponse = HttpRequests.execute(HttpRequest.get(url).build());
+
+    if (httpResponse.getResponseCode() == HttpResponseStatus.NOT_FOUND.getCode()) {
```

Comment:
not found should not result in an IOException. It is an indication that the caller did something wrong, and should be included in the contract of the ArtifactManager.